Feature comments

KosSpb opened the pull request.

KosSpb committed:
Project ExploreWithMe: Comments v0.1.0

<refactor>: переименовать UserResponseDto в UserFullInfoResponseDto;
<refactor>: переименовать InitiatorResponseDto в UserResponseDto;
<feat>: реализовать модель данных Comment, а также DTO-модели CommentFullInfoResponseDto, CommentResponseDto, CommentRequestDto для взаимодействия с контроллерами и передачи данных;
<feat>: реализовать в schema.sql таблицу comments;
<feat>: реализовать логику создания, обновления, удаления и получения комментариев при помощи класса CommentService;
<feat>: реализовать интерфейс DAO CommentRepository для взаимодействия с базой данных;
<feat>: реализовать интерфейс маппера CommentMapper;
<feat>: добавить в DTO-модели EventFullInfoResponseDto и EventResponseDto поле с коллекцией комментариев;
<feat>: добавить в AdminController методы deleteCommentByAdmin() и getFullInfoAboutAllCommentsByAdmin();
<feat>: добавить в PrivateController методы createCommentByUser(), updateCommentByUser(), deleteCommentByUser();

KosSpb committed:
Project ExploreWithMe: Comments v0.1.0

<feat>: добавить в CommentRepository методы findByEventInOrderByIdAsc() и findByEventOrderByIdAsc();
<refactor>: унаследовать класс EventFullInfoResponseDto от EventResponseDto;
<feat>: добавить в DTO-модели EventFullInfoResponseDto от EventResponseDto поле List<CommentResponseDto> comments;
<refactor>: внести изменения в методы класса EventService с учётом добавления в DTO-модели EventFullInfoResponseDto от EventResponseDto поля List<CommentResponseDto> comments для передачи списка комментариев в контроллер;
<refactor>: внести изменение в блок .peek requestedEvents.stream метода getFullInfoAboutAllEventsByAdmin() класса EventService, заменив код в лямбда-выражении на использование метода setViewsAndConfirmedRequests();

KosSpb committed:
Project ExploreWithMe: Comments v0.1.1

<feat>: добавить DTO-модель EventResponseDtoOfComment для использования в CommentFullInfoResponseDto;
<fix>: изменить тип поля event с EventResponseDto на EventResponseDtoOfComment в DTO-модели CommentFullInfoResponseDto;
<refactor>: в сервисных классах сделать отступ от сигнатур методов до кода методов;
<fix>: добавить в модель данных Comment аннотацию @PrePersist с методом defaultIsEdited() для корректного присвоения значений по умолчанию полю isEdited;
<fix>: добавить методу commentToShortDto() интерфейса CommentMapper аннотацию @Mapping с явным указанием полей для маппинга;
<fix>: изменить тип сортировки в методах findByEventInOrderByIdAsc() и findByEventOrderByIdAsc() интерфейса CommentRepository на findByEventInOrderByCreatedAsc() и findByEventOrderByCreatedAsc();
<feat>: добавить в метод updateCommentByUser() класса CommentService проверку на обновление комментария другим пользователем;
<feat>: добавить в метод deleteCommentByUser() класса CommentService проверку на удаление комментария другим пользователем;
<refactor>: изменить реализацию метода getEventResponseDtosWithConfirmedRequests() класса CompilationService на использование мапы из id событий / количества подтверждённых запросов;
<refactor>: переименовать метод setViewsAndConfirmedRequests() класса EventService в setViewsAndConfirmedRequestsAndComments();

KosSpb committed:
Project ExploreWithMe: Comments v0.1.1

<feat>: добавить описание проекта в README.md;

KosSpb committed:
Project ExploreWithMe: Comments v0.1.1

<fix>: обновить feature.json;

alexache commented on core/src/main/java/ru/practicum/explorewithme/controller/AdminController.java:
```diff
@@ -162,4 +161,30 @@ public CompilationResponseDto updateCompilationByAdmin(
         return updatedCompilation;
     }
+
+    @DeleteMapping("/comments/{commentId}")
```
Хорошо бы еще дать админам возможность редактировать пользовательские комментарии, "модерировать"

KosSpb replied:
Добавил.

alexache commented on core/src/main/java/ru/practicum/explorewithme/model/Comment.java:
```diff
@@ -0,0 +1,43 @@
+    private Boolean isEdited;
+    @Column(name = "created_at", nullable = false)
+    @CreationTimestamp
+    private LocalDateTime created;
```
Можно еще добавить время редактирования комментария

alexache commented on core/src/main/java/ru/practicum/explorewithme/service/CommentService.java:
```diff
@@ -0,0 +1,148 @@
+        if (eventToComment.getState() != EventModerationState.PUBLISHED) {
+            throw new ConditionsNotMetException("create comment by user: Cannot comment on an unpublished event");
+        }
+
```
(опционально) можно еще добавить проверку, что юзер посещал событие (через таблица подтвержденных заявок)

KosSpb committed:
Project ExploreWithMe: Comments v0.1.2

<feat>: добавить в AdminController метод updateCommentByAdmin() для получения данных для модерации комментариев пользователей администратором;
<feat>: добавить в CommentService метод updateCommentByAdmin() для редактирования комментариев пользователей администратором;
<refactor>: перенести логику по редактированию комментариев пользователей в метод updateCommentFieldsAndSaveThenReturnDto() класса CommentService;
<feat>: добавить в модель данных Comment поле editedAt, для хранения даты и времени редактирования комментариев;
<feat>: добавить в schema.sql в таблицу comments поле edited_at, для хранения даты и времени редактирования комментариев;
<feat>: добавить тесты на редактирование комментария администратором и получения комментариев с новым полем editedAt в feature.json;

KosSpb committed:
Project ExploreWithMe: Comments v0.1.2

<fix>: обновить feature.json;

alexache commented on core/src/main/java/ru/practicum/explorewithme/service/CommentService.java:
```diff
@@ -145,4 +134,31 @@ public void deleteCommentByAdmin(long commentId) {
+        if (!commentToUpdate.getIsEdited()) {
+            commentToUpdate.setIsEdited(true);
+        }
+        commentToUpdate.setEditedAt(LocalDateTime.now());
```
Время апдейта тоже можно прстоавлять автоматически, это относится к аудит полям и при включенном айдите довольно легко делается еще и версионирование (при необходимости)

KosSpb replied:
То есть можно не устанавливать значение полю editedAt, а просто как-то в БД отслеживать перевод isEdited в true, а уже БД сама timestamp поставит? Правильно понимаю или как-то иначе это реализовать нужно?

alexache commented on core/src/main/java/ru/practicum/explorewithme/service/CommentService.java:
```diff
@@ -145,4 +134,31 @@ public void deleteCommentByAdmin(long commentId) {
+                                                                               Comment commentToUpdate) {
+
+        commentToUpdate.setText(commentRequestDto.getText());
+        if (!commentToUpdate.getIsEdited()) {
```
МОжно просто всегда безусловно ставить в true, никто не пострадает от этого

KosSpb replied:
Исправил.

Александр, спасибо огромное за добрые слова и твои ценные комментарии, сопровождавшие все ревью моих работ на протяжении этого курса.  Я очень тебе благодарен, и рад что у меня был такой отличный ревьюер все эти долгие 11 месяцев. Буду стараться и дальше держать тот уровень качества кода, к которому ты меня приучил, и продолжу развивать свои навыки.
С благодарностью, Константин.
